### ISSUE TITLE
Support multi-view rendering

### ISSUE THREAD

**Issue by dkwingsmt** [flutter/flutter#135530]
In order to make the Flutter engine's rendering system support multiple views, we need to change multiple components across the engine. The plan can be found in design doc [flutter.dev/go/multi-view-pipeline-and-rasterizer](http://flutter.dev/go/multi-view-pipeline-and-rasterizer). Items are listed below. The order of the items does not represent dependency, and some items can be developed in parallel.

- [x] **Item 1:** Eliminate violations of the rendering rule (https://github.com/flutter/engine/pull/45300)
- [x] **Item 2:** Make the pipeline support multiple views. The `Animator` should collect (instead of dispatching) the layer tree in `Render` and dispatch them at the end of `BeginFrame` (https://github.com/flutter/engine/pull/44473).
- [x] **Item 3:** Make the `Rasterizer`'s structure compatible with multiple views. (https://github.com/flutter/engine/pull/45512)
- [x] **Item 4:** Allow the `ExternalViewEmbedder` to handle multiple views. (https://github.com/flutter/engine/pull/51722, https://github.com/flutter/engine/pull/51719, https://github.com/flutter/engine/pull/46169)
- [x] **Abandoned:** Allow the `RasterCache` to handle multiple views. 
  - Per @/stuartmorgan (I can't find the source for now) the raster cache only improves performance on mobile in specific cases, and should not help desktop. The raster cache should be made optional.

**Comment by wanjm** [comment]
what we can do with #45300   and #45512

**Comment by dkwingsmt** [comment]
I'm not very familiar with these issues but are you sure they're related with multi-view?

**Comment by wanjm** [comment]
> I'm not very familiar with these issues but are you sure they're related with multi-view?
with item 1 and item 3 merged, can i do some prototype?
currently I am seriously need the feature of  multi-view and multi-window. so I am trying to do some prototype as multi-view is getting to then end;

it is very difficult for me to compile the engine, as many things can't be downloaded;

**Comment by dkwingsmt** [comment]
Ah, I see. You linked the wrong issues... The two merged PRs are for internal refactor. To make use of them, we also need item 2 and 4, as well as a few more items:
* Embedder API supports multi-view,
* The specific platform embedder (such as Window or macOS) supports multi-view.
* Other features support multi-view (such as https://github.com/flutter/engine/pull/46174 and https://github.com/flutter/engine/pull/46174).

I understand it sounds like a long way to come, and I agree with you. I'm also eager to see the final shape. But multi-view is a feature that touches almost every corner of Flutter, therefore we need to move a step at a time not not to break existing features, as well as designing an API that won't be easily broken for the future.

In short, for a while, compiling custom engine will be the only way to prototype multi-view apps.

**Comment by wanjm** [comment]
will it be ready by the end of this year, as it is in the roadmap  of 2023;
we are now facing the choice of QT or flutter;

**Comment by dkwingsmt** [comment]
We're still on track with that goal. However, we can't promise any progress (including the embedder API) as things can happen, and especially because there aren't many workdays left until the end of the year with the holiday season approaching. I can not say whether Flutter is your best bet either, but here I'll provide as much information as possible.

The current status of the multi-window project is:
* We should be really close to finishing most things before the embedder API.
* We have some agreement with the embedder API but needs more experimentation and reasoning.
* We have a rough idea on the runner API for macOS and Windows but there are some obstacles. 
* Linux support has some serious obstacles and should not be expected in the near future.

And the one thing I can guarantee is that multi-window support is one of our first priorities at the moment, we have a clean plan, and we are dedicated to finishing it. We have some top-priorities customer (project) that needs this feature, so there is no back-off.

Maybe you can start prototyping with a single window at first. There is also a community package (https://pub.dev/packages/desktop_multi_window) that claims to provide multi-window support, although its approach looks sketchy to us and the fact that it works at all feels like a wonder.

That being said, my experience tells me that developing these kind of apps (especially related to desktop) will likely involve customizing the engine or patching some lines (even just to verify) some day. Being unable to compile the engine is a decent risk.

Hope this can help with your decision. :)

**Comment by wanjm** [comment]
thanks

**Comment by dkwingsmt** [comment]
This issue should be considered finished. Basic multi-view rendering is now supported by the embedder API, including adding views, removing views, presenting to views, creating backing store to views, setting view metrics, etc.

**Comment by wanjm** [comment]
when https://github.com/goderbauer/mvp will work without compile the engine;

**Comment by dkwingsmt** [comment]
The embedder API is pretty much finished. So if you have a custom embedder, it should already work. 

We're still working on the macOS and Windows embedders. We have an umbrella issue https://github.com/flutter/flutter/issues/142845.
